tutorials/llm/llama-3/nemo2-sft-peft/nemo2-sft.ipynb

```diff
@@ -0,0 +1,632 @@
+    "from nemo.collections.llm.gpt.data.fine_tuning import FineTuningDataModule\n",
+    "import json\n",
+    "from nemo.utils import logging\n",
+    "import numpy as np\n",
+    "import shutil\n",
+    "\n",
+    "class DollyDataModule(FineTuningDataModule, IOMixin):\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        seq_length: int = 2048,\n",
+    "        tokenizer: Optional[\"TokenizerSpec\"] = None,\n",
+    "        micro_batch_size: int = 4,\n",
+    "        global_batch_size: int = 8,\n",
+    "        rampup_batch_size: Optional[List[int]] = None,\n",
+    "        force_redownload: bool = False,\n",
+    "        delete_raw: bool = True,\n",
+    "        seed: int = 1234,\n",
+    "        memmap_workers: int = 1,\n",
+    "        num_workers: int = 8,\n",
+    "        pin_memory: bool = True,\n",
+    "        persistent_workers: bool = False,\n",
+    "        pad_to_max_length: bool = False,\n",
+    "        packed_sequence_size: int = -1,\n",
+    "    ):\n",
+    "        self.force_redownload = force_redownload\n",
+    "        self.delete_raw = delete_raw\n",
+    "\n",
+    "        super().__init__(\n",
+    "            dataset_root=get_dataset_root(\"dolly\"),\n",
+    "            seq_length=seq_length,\n",
+    "            tokenizer=tokenizer,\n",
+    "            micro_batch_size=micro_batch_size,\n",
+    "            global_batch_size=global_batch_size,\n",
+    "            rampup_batch_size=rampup_batch_size,\n",
+    "            seed=seed,\n",
+    "            memmap_workers=memmap_workers,\n",
+    "            num_workers=num_workers,\n",
+    "            pin_memory=pin_memory,\n",
+    "            persistent_workers=persistent_workers,\n",
+    "            pad_to_max_length=pad_to_max_length,\n",
+    "            packed_sequence_size=packed_sequence_size,\n",
+    "        )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2. Override the `prepare_data` Method\n",
+    "\n",
+    "The `prepare_data` method is responsible for downloading and preprocessing data if needed. If the dataset is already downloaded, you can skip this step.\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def prepare_data(self) -> None:\n",
+    "    # if train file is specified, no need to do anything\n",
+    "    if not self.train_path.exists() or self.force_redownload:\n",
+    "        dset = self._download_data()\n",
+    "        self._preprocess_and_split_data(dset)\n",
+    "    super().prepare_data()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3. Implement Data Download and Preprocessing Logic\n",
+    "\n",
+    "If your dataset requires downloading or preprocessing, implement this logic within helper methods. Skip the download part if it's not needed."
```

Comment:
add definitive article "the"

    "If your dataset requires downloading or preprocessing, implement this logic within the helper methods. Skip the download part if it's not needed."